tests/interlocked_hl_int/interlocked_hl_int.c

```diff
@@ -0,0 +1,287 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+
+#include "testrunnerswitcher.h"
+
+#include "c_pal/threadapi.h"
+#include "macro_utils/macro_utils.h"
+
+#include "c_util/interlocked_hl.h"
+
+TEST_DEFINE_ENUM_TYPE(THREADAPI_RESULT, THREADAPI_RESULT_VALUES);
+TEST_DEFINE_ENUM_TYPE(INTERLOCKED_HL_RESULT, INTERLOCKED_HL_RESULT_VALUES);
+
+BEGIN_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+}
+
+TEST_FUNCTION_INITIALIZE(function_initialize)
+{
+}
+
+TEST_FUNCTION_CLEANUP(function_cleanup)
+{
+}
+
+volatile_atomic int32_t globalValue = 10;
+
+/*
+Tests:
+InterlockedHL_WaitForValue
+InterlockedHL_DecrementAndWake
+*/
+static int decrement_and_wake_helper_thread_function(void* context)
+{
+    (void)context;
+    INTERLOCKED_HL_RESULT result = InterlockedHL_WaitForValue(&globalValue, 9, UINT32_MAX);
+    ASSERT_ARE_EQUAL(INTERLOCKED_HL_RESULT, INTERLOCKED_HL_OK, result);
+    return 0;
+}
+
+TEST_FUNCTION(interlocked_hl_decrement_and_wake_operates_successfully)
+{
+    // + have a helper thread which waits for decremented value
+    // + main thread creates helper thread and decrements value
+    // + helper thread wakes up when the global address value is decremented by 1
+    // arrange
+    globalValue = 10;
+    THREAD_HANDLE helper_thread_handle;
+    THREADAPI_RESULT thread_result = ThreadAPI_Create(&helper_thread_handle, decrement_and_wake_helper_thread_function, NULL);
+    ASSERT_ARE_EQUAL(THREADAPI_RESULT, THREADAPI_OK, thread_result);
+
+    // act
+    //sleep so that the helper thread can go into wait mode
+    ThreadAPI_Sleep(5000);
+    INTERLOCKED_HL_RESULT result = InterlockedHL_DecrementAndWake(&globalValue);
+    ASSERT_ARE_EQUAL(INTERLOCKED_HL_RESULT, INTERLOCKED_HL_OK, result);
+    ASSERT_ARE_EQUAL(INTERLOCKED_HL_RESULT, 9, globalValue);
+
+    // cleanup
+    int return_code;
+    ASSERT_ARE_EQUAL(THREADAPI_RESULT, THREADAPI_OK, ThreadAPI_Join(helper_thread_handle, &return_code));
+    ASSERT_ARE_EQUAL(int, 0, return_code);
+}
+
+/*
+Tests:
+InterlockedHL_WaitForValue
+InterlockedHL_SetAndWakeAll
+*/
+static int set_and_wakeall_helper_thread_function(void* context)
+{
+    (void)context;
+    INTERLOCKED_HL_RESULT result = InterlockedHL_WaitForValue(&globalValue, 15, UINT32_MAX);
+    ASSERT_ARE_EQUAL(INTERLOCKED_HL_RESULT, INTERLOCKED_HL_OK, result);
+    return 0;
+}
+
+TEST_FUNCTION(interlocked_hl_set_and_wakeall_operates_successfully)
+{
+    // + create 10 helper threads which wait on a value
+    // + main thread creates helper threads and sets value for the helper threads to wake up
+    // + main thread joins on the helper threads and ensures that all threads wake up and terminate
+    // arrange
+    globalValue = 10;
+
+    enum { NUMBER_OF_HELPER_THREADS = 10 };

```

Comment:
> enum { NUMBER_OF_HELPER_THREADS = 10 }; [](http://example.com/codeflow?start=3&length=40)

Just wondering would a const size_t do the same? I am genuinely asking so that I know myself.

Reply:
i tried to use const size_t variable. but, the compiler complained when used it for an array as in array[const_variable]. C does not allow it, looks like. so, had to take the enum route.